dbt/adapters/base/impl.py

```diff
@@ -1285,46 +1287,84 @@ def calculate_freshness(
         }
         return adapter_response, freshness
 
+    def calculate_freshness_from_metadata_batch(
+        self,
+        sources: List[BaseRelation],
+        macro_resolver: Optional[MacroResolverProtocol] = None,
+    ) -> Tuple[Optional[AdapterResponse], Dict[BaseRelation, FreshnessResponse]]:
+        # Track schema, identifiers of sources for lookup from batch query
```

Comment:
There's probably a better way to do this. I'm trying to create a mapping ahead of time to use to map the individual rows from the result table back to a source relation, since the result from the warehouse is not guaranteed to be in any particular order (& I've validated that the order is not preserved, even coincidentally).

Main issues I see with this current approach are:
  * source and identifier are technically optional which this (as implemented) doesn't handle
  * there is probably something more reusable that already exists for this purpose, e.g. BaseRelation.matches

I thought about using `BaseRelation.matches` with the per-row `schema` and `identifier` values we get back from the batch query, but my concern there is performance: we'd need to iterate over all inputs to find a matching one, for each result row. Additionally, the casing we get back from the warehouse is pretty out of our control. For example, it seems like snowflake always returns these capitalized, which probably won't match count as an `exact` match using `BaseRelation.matches`. We'd probably need to extend `matches` to optionally return `True` if an approximate match is found. 

Another option is to also expect adapters to implement the `get_relation_last_modified` to include `source.render()` as an `id`, and use that to map from the table result to the. That would require updates to the macro implementations, but only for the adapters that support the MetadataBatch capability, and wouldn't break previous behaviour if such a key was not provided. This is the approach I'm leaning toward currently.

cc @mikealfare -- open to any thoughts or suggestions here!